### PR TITLE
Fix author analytics

### PR DESCRIPTION
**Pull request by Will-Howard**
Previously the data required for the author/post analytics page was fetched from "hybrid views", which consisted of a regular `MATERIALIZED VIEW` in the analytics db which would then be `UNION`-ed with the latest data from the last day or so. This didn't end up working well because:
- The performance was still unreliable, fetching data from just the last day could still be very slow
- Refreshing the materialized views was _extremely_ slow (>3 hours) and consumed a lot of resources in the analytics db
- At some point one of the views started failing to refresh, and , because of it being so slow, this was very difficult to debug

Because of these problems, I've now abandoned this approach and have instead recreated the views as regular collections (in the non-analytics db) which effectively act as materialized views, but where the data is updated incrementally. So we will only have to run a big job to backfill the data once, after which point a cron job will run every hour to update the most recent day of data. The downside of this is that the data is no longer "live", it will be up to an hour behind.

There are a couple of further improvements I think would be good to make in this new setup, for later PRs:
- We could do to improve on the staleness of the data by running the cron job much more frequently, and/or moving from daily to hourly aggregation (this would also make it possible to adjust based on timezones). It will be better to do this once we have an idea of how slow the cron job is to begin with
- Now that the data is in the prod db we could get a better metric for "reads" by comparing to the "reading time" field on the Posts table

For this PR though I have just tried to make it a relatively conservative refactor without changing any of the behaviour.

### Update 2024-01-18

I have now tested this with the prod data (having deployed https://github.com/ForumMagnum/ForumMagnum/pull/8646 ), by running `updateAnalyticsCollections` to backfill about 3 months of data. Summary:
- Loading 1 day of data (which is what the cron job will do in steady state) takes 5-20s in the analytics db, 1-3 seconds in the prod db (to insert it). When it was taking 20s/day for the analytics db it looked like this was due to an expensive cron job running at the same time, once this finished it was much faster
- There was a noticeable increase in CPU usage on the _analytics_ db while this was running, up to about 20% from a 5% baseline (when the "expensive cron job" was running at the same time it was at 70%, but that was mostly the cron job). There was no visible increase on the prod db, which was the main thing I was worried about
- I was also watching out for whether the time to insert into the prod db was increasing as the size of the tables increased (this would be expected because it needs to maintain the indexes). The answer to this is "maybe a bit, but not a lot". It looks like it went up from ~1.5s/day to ~2.5s/day from 0 rows to ~1M rows (this is for the PostViewTimes table, PostViews was similar). When the table is completely full it will be 9M rows, so I don't expect this to be a problem. I think in principle insert time should scale as O(log(N)) (not sure), but even with O(N) it would be fine
- On the read side, I tested some users with a lot of posts (and views) and these queries were no longer the bottleneck, they usually take a few hundred ms. This could be worse with more data but I'm not expecting it to be a problem
- I also found and deleted the bad data that was breaking the current views, so those should start working again now. It would be good to

Given all this, I decided to increase the frequency of the cron job to once every 20 mins, as the load on both dbs will be pretty light.

Running the full backfilling job will take a few hours, so I'll run it overnight tonight (and definitely before merging)

┆Issue is synchronized with this [Asana task](https://app.asana.com/0/1201302964208280/1206352807053101) by [Unito](https://www.unito.io)

### CODE REVIEW

File: packages/lesswrong/server/repos/PostViewTimesRepo.ts [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+    `,
+      [windowStart, windowEnd]
+    );
+
+    // Manually convert the number fields from string to number (as this isn't handled automatically)
+    const typedResults = results.map((views) => ({
+      ...views,
+      viewCount: parseInt(views.viewCount),
+      uniqueViewCount: parseInt(views.uniqueViewCount),
+    }));
```
[the remark this answers is not in the document]
Comment (jpaddison3) [reply packages/lesswrong/server/repos/PostViewTimesRepo.ts:60]: 🪦

**Review by jpaddison3 — APPROVED** [review]
ASTLAC

File: packages/lesswrong/server/repos/PostViewTimesRepo.ts [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+
+  async readsByPost({
+    postIds,
+  }: {
+    postIds: string[];
+  }): Promise<{ postId: string; totalReads: number; }[]> {
+    const results = await this.getRawDb().any<{ postId: string; totalReads: string; }>(`
+      SELECT
+        "postId",
+        sum(CASE WHEN "totalSeconds" >= 30 THEN 1 ELSE 0 END) AS "totalReads"
```
[the remark this answers is not in the document]
Comment (Will-Howard) [reply packages/lesswrong/server/repos/PostViewTimesRepo.ts:134]: _I_ read it ❤️

File: packages/lesswrong/server/analytics/analyticsCron.ts [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+  }
+
+  console.log("Finished updateAnalyticsCollections")
+}
+
+if (isEAForum) {
+  addCronJob({
+    name: "updateAnalyticsCollections",
+    interval: "every 20 minutes",
+    job: async () => updateAnalyticsCollections({}),
```
[the remark this answers is not in the document]
Comment (Will-Howard) [reply packages/lesswrong/server/analytics/analyticsCron.ts:111]: Do you mean why it's defined at the default in the function rather than here? You could do it either way 🤷‍♂️, one reason to prefer this is like "it's simpler if there's no logic in the cron job, so you can keep track of all the logic", but idk

File: packages/lesswrong/server/analytics/analyticsCron.ts [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+async function updateDailyAnalyticsCollection<N extends CollectionNameString>({
+  repo,
+  earliestStartDate,
+  latestEndDate,
+}: {
+  repo: IncrementalViewRepo<N>;
+  earliestStartDate: Date;
+  latestEndDate: Date;
+}) {
+  const { earliestWindowStart, latestWindowEnd } = await repo.getDateBounds()
```
[the remark this answers is not in the document]
Comment (Will-Howard) [reply packages/lesswrong/server/analytics/analyticsCron.ts:19]: In this and below, there is a subtle point that we _do_ always recalculate the most recent day, because it may have had partial data. Agree this comment is useful, I'll just reword it to make sure this is clear

File: packages/lesswrong/server/analytics/analyticsCron.ts [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+  // that are not already covered by the range [`earliestWindowStart`, `latestWindowEnd`)
+  const ranges: {startDate: Date, endDate: Date}[] = []
+
+  let currentDate = moment(earliestStartDate).startOf('day').toDate();
+
+  // Ensure we cover the endDate if it's the same as latestWindowEnd
+  const inclusiveEndDate = latestWindowEnd && moment(latestWindowEnd).subtract(1, 'day').toDate();
+
+  while (currentDate <= latestEndDate) {
+    // Skip dates that are within the [`earliestWindowStart`, `latestWindowEnd`) interval
```
[the remark this answers is not in the document]
Comment (Will-Howard) [reply packages/lesswrong/server/analytics/analyticsCron.ts:32]: As above:
> In this and below, there is a subtle point that we do always recalculate the most recent day, because it may have had partial data. Agree this comment is useful, I'll just reword it to make sure this is clear

File: packages/lesswrong/server/analytics/analyticsCron.ts [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+  ranges.reverse()
+
+  for (const range of ranges) {
+    // Allow individual days to fail
+    try {
+      console.log(`Updating data for range: ${range.startDate.toISOString()} to ${range.endDate.toISOString()}`)
+
+      const data = await repo.calculateDataForDateRange(range);
+
+      console.log(`Calculated ${data.length} rows, upserting data`)
```
[the remark this answers is not in the document]
Comment (Will-Howard) [reply packages/lesswrong/server/analytics/analyticsCron.ts:57]: I thumsed this up but now thinking about it I would like to check the logs on prod at least immediately after launch, so I've made a [followup ticket](https://app.asana.com/0/628521446211730/1206399639116709/f) (deadline 26th Jan) to clean these up

[the remark this answers is not in the document]
Comment (jpaddison3) [reply packages/lesswrong/server/analytics/analyticsCron.ts:57]: You can use a logger on prod! I encourage you to do so.
